File: src/test/java/wile/redstonepen/commands/DemoBuilderTest.java
package wile.redstonepen.commands;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import net.minecraft.core.BlockPos;
import org.junit.jupiter.api.Test;

class DemoBuilderTest
{
  @Test
  void cellOriginIndexZeroReturnsGridOrigin()
  {
    final BlockPos origin = new BlockPos(10, 64, 20);
    assertEquals(origin, DemoBuilder.cellOrigin(origin, 0, 4, 3));
  }

  @Test
  void cellOriginAdvancesAcrossColumnsBeforeAdvancingRow()
  {
    final BlockPos origin = new BlockPos(0, 0, 0);
    // columns=4, spacing=3
    assertEquals(new BlockPos(3, 0, 0), DemoBuilder.cellOrigin(origin, 1, 4, 3));
    assertEquals(new BlockPos(6, 0, 0), DemoBuilder.cellOrigin(origin, 2, 4, 3));
    assertEquals(new BlockPos(9, 0, 0), DemoBuilder.cellOrigin(origin, 3, 4, 3));
    // index 4 wraps to next row
    assertEquals(new BlockPos(0, 0, 3), DemoBuilder.cellOrigin(origin, 4, 4, 3));
    assertEquals(new BlockPos(3, 0, 3), DemoBuilder.cellOrigin(origin, 5, 4, 3));
  }

  @Test
  void cellOriginPreservesYCoordinate()
  {
    final BlockPos origin = new BlockPos(0, 100, 0);
    assertEquals(100, DemoBuilder.cellOrigin(origin, 7, 3, 5).getY());
  }

  @Test
  void cellOriginRejectsZeroColumns()
  {
    assertThrows(IllegalArgumentException.class,
      () -> DemoBuilder.cellOrigin(BlockPos.ZERO, 0, 0, 1));
  }

  @Test
  void cellOriginRejectsZeroSpacing()
  {
    assertThrows(IllegalArgumentException.class,
      () -> DemoBuilder.cellOrigin(BlockPos.ZERO, 0, 4, 0));
  }

  @Test
  void cellOriginRejectsNegativeIndex()
  {
    assertThrows(IllegalArgumentException.class,
      () -> DemoBuilder.cellOrigin(BlockPos.ZERO, -1, 4, 3));
  }

  @Test
  void regionVolumeIsInclusiveOnBothEnds()
  {
    assertEquals(1, DemoBuilder.regionVolume(BlockPos.ZERO, BlockPos.ZERO));
    assertEquals(8, DemoBuilder.regionVolume(BlockPos.ZERO, new BlockPos(1, 1, 1)));
    assertEquals(27, DemoBuilder.regionVolume(BlockPos.ZERO, new BlockPos(2, 2, 2)));
  }

  @Test
  void regionVolumeOrderIndependent()
  {
    final BlockPos a = new BlockPos(5, 5, 5);
    final BlockPos b = new BlockPos(0, 0, 0);
    assertEquals(DemoBuilder.regionVolume(a, b), DemoBuilder.regionVolume(b, a));
  }
}

File: common/src/test/kotlin/wile/redstonepen/commands/DemoBuilderTest.kt
package wile.redstonepen.commands

import net.minecraft.core.BlockPos
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class DemoBuilderTest {
    @Test
    fun cellOriginIndexZeroReturnsGridOrigin() {
        val origin = BlockPos(10, 64, 20)
        assertEquals(origin, DemoBuilder.cellOrigin(origin, 0, 4, 3))
    }

    @Test
    fun cellOriginAdvancesAcrossColumnsBeforeAdvancingRow() {
        val origin = BlockPos(0, 0, 0)
        assertEquals(BlockPos(3, 0, 0), DemoBuilder.cellOrigin(origin, 1, 4, 3))
        assertEquals(BlockPos(6, 0, 0), DemoBuilder.cellOrigin(origin, 2, 4, 3))
        assertEquals(BlockPos(9, 0, 0), DemoBuilder.cellOrigin(origin, 3, 4, 3))
        assertEquals(BlockPos(0, 0, 3), DemoBuilder.cellOrigin(origin, 4, 4, 3))
        assertEquals(BlockPos(3, 0, 3), DemoBuilder.cellOrigin(origin, 5, 4, 3))
    }

    @Test
    fun cellOriginPreservesYCoordinate() {
        val origin = BlockPos(0, 100, 0)
        assertEquals(100, DemoBuilder.cellOrigin(origin, 7, 3, 5).y)
    }

    @Test
    fun cellOriginRejectsZeroColumns() {
        assertThrows<IllegalArgumentException> { DemoBuilder.cellOrigin(BlockPos.ZERO, 0, 0, 1) }
    }

    @Test
    fun cellOriginRejectsZeroSpacing() {
        assertThrows<IllegalArgumentException> { DemoBuilder.cellOrigin(BlockPos.ZERO, 0, 4, 0) }
    }

    @Test
    fun cellOriginRejectsNegativeIndex() {
        assertThrows<IllegalArgumentException> { DemoBuilder.cellOrigin(BlockPos.ZERO, -1, 4, 3) }
    }

    @Test
    fun regionVolumeIsInclusiveOnBothEnds() {
        assertEquals(1, DemoBuilder.regionVolume(BlockPos.ZERO, BlockPos.ZERO))
        assertEquals(8, DemoBuilder.regionVolume(BlockPos.ZERO, BlockPos(1, 1, 1)))
        assertEquals(27, DemoBuilder.regionVolume(BlockPos.ZERO, BlockPos(2, 2, 2)))
    }

    @Test
    fun regionVolumeOrderIndependent() {
        val a = BlockPos(5, 5, 5)
        val b = BlockPos(0, 0, 0)
        assertEquals(DemoBuilder.regionVolume(a, b), DemoBuilder.regionVolume(b, a))
    }
}
